### DISCUSSION TITLE
Do we expire keys if cached is enabled in CachedEnforcer ?

### DISCUSSION THREAD

**Discussion by yugant007** [casbin/casbin#1199] [Q&A] [answered]
I am using CachedEnforcer, i have enabled cached by calling `EnableCache(true)` method. Also i have set key TTL as `SetExpireTime(10)`. But when deep dive in casbin code, i don't see keys getting expired. I am using golang.

**Comment by hsluoyz** [comment]
Please paste full code here

**Comment by yugant007** [comment]
This is how i create casbin cached enforcer instance and intialized it with expire time and enabled cache.

`
enforcer, err := casbin.NewCachedEnforcer(configPath, RBACLenderDashboardSvc.Adapter)

	if err != nil {

		return err

	}

	enforcer.EnableCache(true)

	enforcer.SetExpireTime(10)

`

First question this expire time passed in `SetExpireTime` method is in sec or millisecond ?

Below is cached enforcer authorisation method present in casbin library

 `
func (e *CachedEnforcer) Enforce(rvals ...interface{}) (bool, error) {

	if atomic.LoadInt32(&e.enableCache) == 0 {

		return e.Enforcer.Enforce(rvals...)

	}

	key, ok := e.getKey(rvals...)

	if !ok {

		return e.Enforcer.Enforce(rvals...)

	}


	if res, err := e.getCachedResult(key); err == nil {
		return res, nil
	} else if err != cache.ErrNoSuchKey {
		return res, err
	}

	res, err := e.Enforcer.Enforce(rvals...)
	if err != nil {
		return false, err
	}

	err = e.setCachedResult(key, res, e.expireTime)
	return res, err
}
`

Implementation of `setCachedResult` method in casbin library 

`
func (e *CachedEnforcer) setCachedResult(key string, res bool, extra ...interface{}) error {

	e.locker.Lock()

	defer e.locker.Unlock()

	return e.cache.Set(key, res, extra...)

}
`

Implementation of `Set` method in casbin library 

`
func (c *DefaultCache) Set(key string, value bool, extra ...interface{}) error {

	(*c)[key] = value

	return nil

}
`

So we see we are not using expire time in set method ?

**Comment by hsluoyz** [comment]
@yugant007 plz create a github issue

**Reply by yugant007** [reply]
Created https://github.com/casbin/casbin/issues/1202

**Reply by hsluoyz** [reply, marked as answer]
bug confirmed and fixed in: https://github.com/casbin/casbin/pull/1203
